src/core_modules/capture-core/components/WidgetProfile/DataEntry/factory/DataElementFactory.js

```diff
@@ -0,0 +1,196 @@
+} from '../../../../storageControllers/cache.types';
+import {
+    DataElement,
+    DateDataElement,
+    DataElementUnique,
+    dataElementUniqueScope,
+    dataElementTypes,
+} from '../../../../metaData';
+import { OptionSetFactory } from './OptionSetFactory';
+import { convertFormToClient, convertClientToServer } from '../../../../converters';
+import { getApi } from '../../../../d2/d2Instance';
+
+type ConstructorInput = {|
+    cachedTrackedEntityAttributes: Map<string, CachedTrackedEntityAttribute>,
+    cachedOptionSets: Map<string, CachedOptionSet>,
+|};
+
+export class DataElementFactory {
+    static errorMessages = {
+        TRACKED_ENTITY_ATTRIBUTE_NOT_FOUND:
+            'TrackedEntityAttributeId missing from programTrackedEntityAttribute or trackedEntityAttribute not found',
+    };
+
+    static buildtetFeatureType(featureType: 'POINT' | 'POLYGON') {
+        const dataElement = new DataElement((o) => {
+            o.id = `FEATURETYPE_${featureType}`;
+            o.name = featureType === 'POINT' ? i18n.t('Coordinate') : i18n.t('Area');
+            o.formName = o.name;
+            o.compulsory = false;
+            o.displayInForms = true;
+            o.disabled = false;
+            o.type = featureType === 'POINT' ? dataElementTypes.COORDINATE : dataElementTypes.POLYGON;
+        });
+        return dataElement;
+    }
+
+    static _buildDataElementUnique(
+        dataElement: DataElement,
+        cachedTrackedEntityAttribute: CachedTrackedEntityAttribute,
+    ) {
+        return new DataElementUnique((o) => {
+            o.scope = cachedTrackedEntityAttribute.orgunitScope
+                ? dataElementUniqueScope.ORGANISATION_UNIT
+                : dataElementUniqueScope.ENTIRE_SYSTEM;
+
+            o.onValidate = (value: any, contextProps: Object = {}) => {
+                const serverValue = pipe(convertFormToClient, convertClientToServer)(
+                    value,
+                    cachedTrackedEntityAttribute.valueType,
+                );
+
+                if (contextProps.onGetUnsavedAttributeValues) {
+                    const unsavedAttributeValues = contextProps.onGetUnsavedAttributeValues(dataElement.id);
+                    if (unsavedAttributeValues) {
+                        const foundValue = unsavedAttributeValues.find(usav => usav === serverValue);
+                        if (foundValue) {
+                            return {
+                                valid: false,
+                                data: {
+                                    attributeValueExistsUnsaved: true,
+                                },
+                            };
+                        }
+                    }
+                }
+
+                let requestPromise;
+                if (o.scope === dataElementUniqueScope.ORGANISATION_UNIT) {
+                    const orgUnitId = contextProps.orgUnitId;
+                    requestPromise = getApi().get('trackedEntityInstances', {
```

Comment:
I thought we are moving away from `getApi`

Reply:
Hey, 
I was not aware we are moving away from `getApi`. What are we supposed to use instead? 
In this case, I cannot use the `useDataQuery` hook from `@dhis2/app-runtime` because I am not inside a react component. What do you suggest to use insted of `getApi`?
Thank you!